checker/src/org/checkerframework/checker/index/searchindex/SearchIndexAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,152 @@
+import org.checkerframework.checker.index.qual.SearchIndexFor;
+import org.checkerframework.checker.index.qual.SearchIndexUnknown;
+import org.checkerframework.common.basetype.BaseAnnotatedTypeFactory;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.value.ValueAnnotatedTypeFactory;
+import org.checkerframework.common.value.ValueChecker;
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.framework.util.AnnotationBuilder;
+import org.checkerframework.framework.util.MultiGraphQualifierHierarchy;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+/**
+ * The Search Index Checker is used to help type the results of calls to the JDK's binary search
+ * methods. It is part of the Index Checker.
+ */
+public class SearchIndexAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    public final AnnotationMirror UNKNOWN, BOTTOM;
+
+    public SearchIndexAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+        UNKNOWN = AnnotationUtils.fromClass(elements, SearchIndexUnknown.class);
+        BOTTOM = AnnotationUtils.fromClass(elements, SearchIndexBottom.class);
+        this.postInit();
+    }
+
+    /**
+     * Provides a way to query the Constant Value Checker, which computes the values of expressions
+     * known at compile time (constant propagation and folding).
+     */
+    ValueAnnotatedTypeFactory getValueAnnotatedTypeFactory() {
+        return getTypeFactoryOfSubchecker(ValueChecker.class);
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        return new LinkedHashSet<>(
+                Arrays.asList(
+                        SearchIndexFor.class,
+                        SearchIndexBottom.class,
+                        SearchIndexUnknown.class,
+                        NegativeIndexFor.class));
+    }
+
+    @Override
+    public QualifierHierarchy createQualifierHierarchy(
+            MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+        return new SearchIndexQualifierHierarchy(factory);
+    }
+
+    private final class SearchIndexQualifierHierarchy extends MultiGraphQualifierHierarchy {
+
+        public SearchIndexQualifierHierarchy(
+                MultiGraphQualifierHierarchy.MultiGraphFactory factory) {
+            super(factory);
+        }
+
+        @Override
+        public AnnotationMirror greatestLowerBound(AnnotationMirror a1, AnnotationMirror a2) {
+            if (AnnotationUtils.areSame(a1, UNKNOWN)) {
+                return a2;
+            }
+            if (AnnotationUtils.areSame(a2, UNKNOWN)) {
+                return a1;
+            }
+            if (AnnotationUtils.areSame(a1, BOTTOM)) {
+                return a1;
+            }
+            if (AnnotationUtils.areSame(a2, BOTTOM)) {
+                return a2;
+            }
+            if (isSubtype(a1, a2)) {
```

Comment:
This isn't right.  The GLB of `@SearchIndex({"a", "b"})` and `@NegativeIndexFor("a")` is `@NegativeIndexFor({"a", "b"})`.